### PR TITLE
gh-87092: expose the compiler's codegen to python for unit tests

### PR DESCRIPTION
**Pull request by iritkatriel**
With this we can run just codegen from python, and get the un-optimized instruction sequence for an AST.

* Issue: gh-87092

### CODE REVIEW

File: Include/internal/pycore_compile.h [hunk 1]
```diff
@@ -40,6 +40,13 @@ extern int _PyAST_Optimize(
     _PyASTOptimizeState *state);
 
 /* Access compiler internals for unit testing */
+
+PyAPI_FUNC(PyObject*) _PyCompile_CodeGen(
```
Comment (markshannon) [inline Include/internal/pycore_compile.h:44]: Does this (or the other functions) need to be an API function?

File: Lib/test/support/bytecode_helper.py [hunk 2]
```diff
@@ -58,24 +57,35 @@ def Label(self):
...
-                instructions.append(item)
-            else:
-                assert isinstance(item, tuple)
-                inst = list(reversed(item))
-                opcode = dis.opmap[inst.pop()]
-                oparg = inst.pop() if opcode in self.HAS_ARG_OR_TARGET else 0
-                loc = inst + [-1] * (4 - len(inst))
-                instructions.append((opcode, oparg, *loc))
-        return instructions
+    def compareInstructions(self, actual_, expected_):
```
Comment (markshannon) [inline Lib/test/support/bytecode_helper.py]: Does this compare, or assert that they are equal? `assertInstructionsEqual` perhaps?

File: Python/compile.c [hunk 3]
```diff
@@ -603,43 +603,52 @@ compiler_init(struct compiler *c)
...
+{
+    struct compiler c;
+    if (!compiler_init(&c)) {
+        return NULL;
+    }
+
+    PyCompilerFlags flags = pflags ? *pflags : _PyCompilerFlags_INIT;
+
+    PyCodeObject *co = NULL;
+    if (!compiler_setup(&c, mod, filename, flags, optimize, arena)) {
```
Comment (markshannon) [inline Python/compile.c]: Why is there a `compiler_init` and a `compiler_setup`?
Given it needs to be freed anyway, would it makes sense to combine into a `new_compiler()` rather than have partially initialized compilers on the stack?
```
PyCodeObject *
_PyAST_Compile(mod_ty mod, PyObject *filename, PyCompilerFlags *pflags,
               int optimize, PyArena *arena)
{
    PyCompilerFlags flags = pflags ? *pflags : _PyCompilerFlags_INIT;
    struct compiler *c = new_compiler(mod, filename, flags, optimize, arena);
    if (c == NULL) {
        return NULL;
    }
    co = compiler_mod(c, mod);
    compiler_free(c);
    return co;
}
```

File: Python/compile.c [hunk 4]
```diff
@@ -9894,7 +9913,10 @@ cfg_to_instructions(cfg_builder *g)
         for (int i = 0; i < b->b_iused; i++) {
             struct instr *instr = &b->b_instr[i];
             location loc = instr->i_loc;
-            int arg = HAS_TARGET(instr->i_opcode) ? instr->i_target->b_label : instr->i_oparg;
+            int arg = instr->i_oparg;
```
Comment (markshannon) [inline Python/compile.c]: Why this change? Or, why did this work before?
If `HAS_TARGET(instr->i_opcode) && instr->i_target == NULL` can be true, this would have segfaulted.

**Review by markshannon — COMMENTED** [review]
This is a nice improvement.
I have a few questions and comments.

File: Include/internal/pycore_compile.h (hunk 1 above)
Comment (iritkatriel) [reply Include/internal/pycore_compile.h:44]: Can _testinternalcapi import them otherwise?

File: Lib/test/support/bytecode_helper.py (hunk 2 above)
Comment (iritkatriel) [reply Lib/test/support/bytecode_helper.py]: assertInstructionsMatch perhaps (there is some processing, it's not exactly equality).

File: Include/internal/pycore_compile.h (hunk 1 above)
Comment (markshannon) [reply Include/internal/pycore_compile.h:44]: I think so, that why it has "internal" in the name.
It defines `Py_BUILD_CORE_BUILTIN`

File: Python/compile.c (hunk 4 above)
Comment (iritkatriel) [reply Python/compile.c]: It didn't need to change, I'll revert.

File: Include/internal/pycore_compile.h (hunk 1 above)
Comment (iritkatriel) [reply Include/internal/pycore_compile.h:44]: It doesn't seem to work:

`[ERROR] _testinternalcapi failed to import: dlopen(/Users/iritkatriel/src/cpython-654/build/lib.macosx-13.0-x86_64-3.12-pydebug/_testinternalcapi.cpython-312d-darwin.so, 0x0002): symbol not found in flat namespace '__PyCompile_CodeGen'`

Comment (iritkatriel) [reply Include/internal/pycore_compile.h:44]: I'll merge this PR now, it's getting into merge conflicts all the time. We can remove these from the API later if it's possible.
